File: apps/worker/src/lib/mistral-ocr.ts
import { getMistralApiBaseUrl, getMistralApiKey } from "@aqshara/config";
import {
  isR2ObjectStorageConfigured,
  presignGetSourceObject,
} from "@aqshara/storage";

export type MistralOcrResult = {
  pages: Array<{
    index: number;
    markdown: string;
  }>;
  text: string;
};

/**
 * Runs Mistral Document OCR. Prefers a presigned R2 URL when configured;
 * otherwise skips OCR (caller should treat as failure or no-op).
 */
export async function runMistralOcrOnPdfKey(input: {
  storageKey: string;
  /** 0-based page indices to send to the API; omit for full document. */
  pages?: number[];
}): Promise<MistralOcrResult | null> {
  const apiKey = getMistralApiKey();
  if (!apiKey) {
    return null;
  }

  let documentUrl: string | undefined;
  if (isR2ObjectStorageConfigured()) {
    documentUrl = await presignGetSourceObject({
      key: input.storageKey,
      expiresSeconds: 900,
    });
  }

  if (!documentUrl) {
    return null;
  }

  const base = getMistralApiBaseUrl().replace(/\/$/, "");
  const body: Record<string, unknown> = {
    model: "mistral-ocr-latest",
    document: {
      type: "document_url",
      document_url: documentUrl,
    },
  };

  if (input.pages && input.pages.length > 0) {
    body.pages = input.pages;
  }

  const res = await fetch(`${base}/v1/ocr`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const errText = await res.text().catch(() => "");
    throw new Error(
      `Mistral OCR failed: ${res.status} ${errText.slice(0, 500)}`,
    );
  }

  const json = (await res.json()) as {
    pages?: Array<{ markdown?: string; index?: number }>;
  };

  const pages =
    json.pages
      ?.filter(
        (page): page is { markdown: string; index: number } =>
          typeof page.index === "number" &&
          typeof page.markdown === "string" &&
          page.markdown.length > 0,
      )
      .map((page) => ({
        index: page.index,
        markdown: page.markdown,
      })) ?? [];

  return {
    pages,
    text: pages.map((page) => page.markdown).join("\n\n"),
  };
}

File: apps/worker/src/worker/jobs/mistral_ocr.py
from __future__ import annotations

from typing import Any

import requests

from worker.config import Settings
from worker.storage import StorageService


class MistralOcrClient:
    def __init__(self, settings: Settings, storage: StorageService) -> None:
        self._settings = settings
        self._storage = storage

    def run(self, *, storage_key: str, pages: list[int] | None = None) -> dict[str, Any] | None:
        if not self._settings.mistral_enabled:
            return None
        document_url = self._storage.presign_source_get(storage_key)
        body: dict[str, Any] = {
            "model": "mistral-ocr-latest",
            "document": {
                "type": "document_url",
                "document_url": document_url,
            },
        }
        if pages:
            body["pages"] = pages
        response = requests.post(
            f"{str(self._settings.mistral_api_base_url).rstrip('/')}/v1/ocr",
            headers={
                "Authorization": f"Bearer {self._settings.mistral_api_key}",
                "Content-Type": "application/json",
            },
            json=body,
            timeout=self._settings.mistral_ocr_timeout_ms / 1000,
        )
        if not response.ok:
            raise RuntimeError(
                f"Mistral OCR failed: {response.status_code} {response.text[:500]}"
            )
        payload = response.json()
        pages_out = [
            {"index": page["index"], "markdown": page["markdown"]}
            for page in payload.get("pages", [])
            if isinstance(page.get("index"), int) and isinstance(page.get("markdown"), str) and page["markdown"]
        ]
        return {"pages": pages_out, "text": "\n\n".join(page["markdown"] for page in pages_out)}
